information_for_beamlines_ds/information_for_beamlines.py

```diff
@@ -0,0 +1,173 @@
+        unit='A',
+        format='%8.6f',
+        access=AttrWriteType.READ,
+        display_level=DispLevel.EXPERT,
+        doc="Forwarded beam current from BIM."
+    )
+
+    BIMState = proxy_attribute(
+        dtype="DevState",
+        property_name="BIMStateTag",
+        access=AttrWriteType.READ,
+        display_level=DispLevel.EXPERT,
+        doc="State of the beam intensity monitor."
+    )
+
+    MPSState = proxy_attribute(
+        dtype="DevState",
+        property_name="MPSStateTag",
+        access=AttrWriteType.READ,
+        display_level=DispLevel.EXPERT,
+        doc="State of the MPS PLC device."
+    )
+
+    # ----------
+    # Logical attributes
+    # ----------
+
+    @logical_attribute(
+        dtype='str',
+        label="Operation Status",
+        bind=["BeamCurrent","ExperimentEnable","InjectionStatus"],
+        doc="One of: Injection, MDT (Machine Dedicated Time), Experiment Enable.",)
+    def OperationStatus(self, beam_c,exp_en,inj_stat):
+        if inj_stat:
+            return 'Injection'
+        elif beam_c > 1 and exp_en:
+            return 'Experiment Enable'
+        else:
+            return 'MDT'
+
+    @logical_attribute(
+        dtype=float,
+        unit='mA',
+        bind=["BeamCurrent"],
+        format='%4.3f',
+        label='Beam Current (mA)',
+        doc='Beam current in miliamperes.',
+        display_level=DispLevel.EXPERT)
+    def BeamCurrent_mA(self, data):
+        current_mA = data * 1000.0
+        # :TODO: Clarify if it's possible to push AttrQuality here.
+        # if abs(current_mA) >= 1e-6:
+        #     return current_mA , time(), AttrQuality.ATTR_VALID
+        # else:
+        #     return current_mA , time(), AttrQuality.ATTR_INVALID
+        return current_mA
+
+    def always_executed_hook(self):
+        pass
+
+    def delete_device(self):
+        pass
+
+    def state_from_data(self, data):
+        mps = data['MPSState']
+        bim = data['BIMState']
+        if mps == DevState.FAULT:
+            self.set_status('MPS PLC device is in FAULT.')
+            return DevState.FAULT
+        elif mps != DevState.ON and mps != DevState.RUNNING:
+            self.set_status('MPS PLC device is in an unexpected state: %s' % str(mps))
+            return DevState.ALARM
+        if bim == DevState.FAULT:
+            self.set_status('BIM device is in FAULT.')
+            return DevState.FAULT
+        elif bim != DevState.ON and bim != DevState.RUNNING:
+            self.set_status('BIM device is in an unexpected state: %s' % str(bim))
+            return DevState.ALARM
+
+        self.set_state(DevState.ON)
+        self.set_status('Everything is OK.')
+
+        # ------------------
+        # Attributes methods
+        # ------------------
+
+    def read_GeneralInfo(self):
+        return self.general_info
+
+    def write_GeneralInfo(self, value):
+        self.general_info = value
+        # self.push_change_event('GeneralInfo')
```

Comment:
Please uncomment those lines and find a way to configure pushing events for this attribute from the code.
This also applies to the `FillingPattern` attribute.